File: riveredge-backend/src/apps/kuaizhizao/schemas/computation_config.py
"""
需求计算参数配置Schema

提供需求计算参数配置相关的数据验证Schema。

Author: Luigi Lu
Date: 2025-01-14
"""

from typing import Optional, Dict, Any
from datetime import datetime
from pydantic import BaseModel, Field, field_validator


class ComputationConfigBase(BaseModel):
    """需求计算参数配置基础Schema"""
    config_code: str = Field(..., max_length=50, description="配置编码")
    config_name: str = Field(..., max_length=200, description="配置名称")
    
    config_scope: str = Field("global", max_length=50, description="配置维度（global/material/warehouse/material_warehouse）")
    material_id: Optional[int] = Field(None, description="物料ID")
    material_code: Optional[str] = Field(None, max_length=50, description="物料编码")
    material_name: Optional[str] = Field(None, max_length=200, description="物料名称")
    warehouse_id: Optional[int] = Field(None, description="仓库ID")
    warehouse_code: Optional[str] = Field(None, max_length=50, description="仓库编码")
    warehouse_name: Optional[str] = Field(None, max_length=200, description="仓库名称")
    
    computation_params: Dict[str, Any] = Field(..., description="计算参数（JSON格式）")
    
    is_template: bool = Field(False, description="是否为模板")
    template_name: Optional[str] = Field(None, max_length=200, description="模板名称")
    
    is_active: bool = Field(True, description="是否启用")
    priority: int = Field(0, description="优先级")
    description: Optional[str] = Field(None, description="配置描述")
    
    @field_validator("config_scope")
    def validate_config_scope(cls, v):
        """验证配置维度"""
        if v not in ["global", "material", "warehouse", "material_warehouse"]:
            raise ValueError("配置维度必须是global、material、warehouse或material_warehouse")
        return v


class ComputationConfigCreate(ComputationConfigBase):
    """创建需求计算参数配置Schema"""
    pass


class ComputationConfigUpdate(BaseModel):
    """更新需求计算参数配置Schema"""
    config_name: Optional[str] = Field(None, max_length=200, description="配置名称")
    computation_params: Optional[Dict[str, Any]] = Field(None, description="计算参数（JSON格式）")
    is_template: Optional[bool] = Field(None, description="是否为模板")
    template_name: Optional[str] = Field(None, max_length=200, description="模板名称")
    is_active: Optional[bool] = Field(None, description="是否启用")
    priority: Optional[int] = Field(None, description="优先级")
    description: Optional[str] = Field(None, description="配置描述")


class ComputationConfigResponse(ComputationConfigBase):
    """需求计算参数配置响应Schema"""
    id: int
    uuid: str
    tenant_id: int
    created_at: datetime
    updated_at: datetime
    created_by: Optional[int]
    updated_by: Optional[int]
    
    class Config:
        from_attributes = True


class ComputationConfigListResponse(BaseModel):
    """需求计算参数配置列表响应Schema"""
    data: list[ComputationConfigResponse]
    total: int
    success: bool = True

File: riveredge-frontend/src/apps/kuaizhizao/services/computation-config.ts
/**
 * 需求计算参数配置API服务
 *
 * 提供需求计算参数配置相关的API调用接口
 *
 * @author Luigi Lu
 * @date 2025-01-14
 */

import { apiRequest } from '../../../services/api';

/**
 * 计算参数配置接口定义
 */
export interface ComputationConfig {
  id?: number;
  uuid?: string;
  tenant_id?: number;
  config_code?: string;
  config_name?: string;
  config_scope?: 'global' | 'material' | 'warehouse' | 'material_warehouse';
  material_id?: number;
  material_code?: string;
  material_name?: string;
  warehouse_id?: number;
  warehouse_code?: string;
  warehouse_name?: string;
  computation_params?: Record<string, any>;
  is_template?: boolean;
  template_name?: string;
  is_active?: boolean;
  priority?: number;
  description?: string;
  created_at?: string;
  updated_at?: string;
  created_by?: number;
  updated_by?: number;
}

/**
 * 计算参数配置列表参数
 */
export interface ComputationConfigListParams {
  skip?: number;
  limit?: number;
  config_scope?: 'global' | 'material' | 'warehouse' | 'material_warehouse';
  material_id?: number;
  warehouse_id?: number;
  is_template?: boolean;
  is_active?: boolean;
  keyword?: string;
}

/**
 * 计算参数配置列表响应
 */
export interface ComputationConfigListResponse {
  data: ComputationConfig[];
  total: number;
  success: boolean;
}

/**
 * 创建计算参数配置
 */
export async function createComputationConfig(data: Partial<ComputationConfig>): Promise<ComputationConfig> {
  return apiRequest<ComputationConfig>({
    url: '/apps/kuaizhizao/computation-configs',
    method: 'POST',
    data,
  });
}

/**
 * 获取计算参数配置列表
 */
export async function listComputationConfigs(params?: ComputationConfigListParams): Promise<ComputationConfigListResponse> {
  return apiRequest<ComputationConfigListResponse>({
    url: '/apps/kuaizhizao/computation-configs',
    method: 'GET',
    params,
  });
}

/**
 * 获取计算参数配置详情
 */
export async function getComputationConfig(id: number): Promise<ComputationConfig> {
  return apiRequest<ComputationConfig>({
    url: `/apps/kuaizhizao/computation-configs/${id}`,
    method: 'GET',
  });
}

/**
 * 更新计算参数配置
 */
export async function updateComputationConfig(id: number, data: Partial<ComputationConfig>): Promise<ComputationConfig> {
  return apiRequest<ComputationConfig>({
    url: `/apps/kuaizhizao/computation-configs/${id}`,
    method: 'PUT',
    data,
  });
}

/**
 * 删除计算参数配置
 */
export async function deleteComputationConfig(id: number): Promise<void> {
  return apiRequest<void>({
    url: `/apps/kuaizhizao/computation-configs/${id}`,
    method: 'DELETE',
  });
}

/**
 * 获取用于计算的参数配置
 */
export interface ComputationParamsResponse {
  success: boolean;
  params: Record<string, any>;
}

export async function getComputationParams(materialId?: number, warehouseId?: number): Promise<ComputationParamsResponse> {
  return apiRequest<ComputationParamsResponse>({
    url: '/apps/kuaizhizao/computation-configs/for-computation/params',
    method: 'GET',
    params: {
      material_id: materialId,
      warehouse_id: warehouseId,
    },
  });
}
